src/ray/common/status_or.h

```diff
@@ -0,0 +1,257 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <stdexcept>
+#include <string_view>
+#include <type_traits>
+#include <utility>
+
+#include "absl/base/attributes.h"
+#include "ray/common/status.h"
+
+namespace ray {
+
+// TODO(hjiang): Introduce statusor related macros.
+template <typename T>
+class StatusOr {
+ public:
+  StatusOr() noexcept = default;
+  // NOLINTNEXTLINE(runtime/explicit)
+  StatusOr(Status status) : status_(std::move(status)) {}
+  // NOLINTNEXTLINE(runtime/explicit)
+  StatusOr(const T &data) : status_(Status::OK()), data_(data) {}
+  // NOLINTNEXTLINE(runtime/explicit)
+  StatusOr(T &&data) : status_(Status::OK()), data_(std::move(data)) {}
+
+  template <typename... Args>
+  explicit StatusOr(std::in_place_t ip, Args &&...args)
+      : data_(std::forward<Args>(args)...) {}
+
+  template <typename U>
+  StatusOr(const StatusOr<U> &status_or) {
+    if (status_or.ok()) {
+      data_ = status_or.value();
+      status_ = Status::OK();
+    } else {
+      status_ = status_or.status();
+    }
+  }
+
+  template <typename U>
+  StatusOr(StatusOr<U> &&status_or) {
+    if (status_or.ok()) {
+      data_ = std::move(status_or).value();
+      status_ = Status::OK();
+    } else {
+      status_ = std::move(status_or).status();
+    }
+  }
+
+  StatusOr(const StatusOr &) = default;
+  StatusOr &operator=(const StatusOr &) = default;
+
+  StatusOr(StatusOr &&) noexcept(std::is_nothrow_move_constructible_v<T>) = default;
+  StatusOr &operator=(StatusOr &&) noexcept(std::is_nothrow_move_assignable_v<T>) =
+      default;
+
+  ~StatusOr() noexcept(std::is_nothrow_destructible_v<T>) = default;
+
+  // Returns whether or not this `ray::StatusOr<T>` holds a `T` value. This
+  // member function is analogous to `ray::Status::ok()` and should be used
+  // similarly to check the status of return values.
+  //
+  // Example:
+  //
+  // StatusOr<Foo> result = DoBigCalculationThatCouldFail();
+  // if (result.ok()) {
+  //    // Handle result
+  // else {
+  //    // Handle error
+  // }
+  bool ok() const { return status_.ok(); }
+  bool has_value() const { return ok(); }
+  explicit operator bool() const { return ok(); }
+
+  template <typename U>
+  T value_or(U &&u) {
+    return ok() ? get() : T{std::forward<U>(u)};
+  }
+
+  ABSL_MUST_USE_RESULT StatusCode code() const { return status_.code(); }
+
+  ABSL_MUST_USE_RESULT std::string message() const { return status_.message(); }
+
+  // Returns a reference to the current `ray::Status` contained within the
+  // `ray::StatusOr<T>`. If `ray::StatusOr<T>` contains a `T`, then this
+  // function returns `ray::Ok()`.
+  ABSL_MUST_USE_RESULT const Status &status() const & { return status_; }
+  ABSL_MUST_USE_RESULT Status status() && {
+    Status new_status = std::move(status_);
```

Comment:
I think Ray Status doesn't have a move constructor

Reply:
That's funny... Two things:
- I could followup with another PR and check if there're any missing function for status;
- As I mentioned always, it's never a bad practice to use move to indicate ownership transfer, even if it's copy semantics underlying.